enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,375 @@
+    // served on the route.
+    RouteCertificateDefaultType RouteIngressConditionType = "Default"
+
+    // RouteCertificateCustomType denotea that there is a custom certificate 
+    // being served on the route. 
+    RouteCertificateCustomType RouteIngressConditionType = "Custom"
+
+    // RouteCertificateManagedType denotes that the annotation is in use to 
+    // inject a third-party managed secret containing the certificate that 
+    // is being served on the route.
+    RouteCertificateManagedType RouteIngressConditionType = "Managed"
+
+)
+
+```
+
+#### Variation
+
+##### Alternative to the tls-secret-reference annotation
+
+***Note***: This idea has been dropped in favor of a secret reference through an annotation.
+
+The alternative to the annotation was to provide a new API field in the Route API,
+through which the user could provide the certificate reference.
+
+```go
+type CertificateReference struct {
+	// certificate resource name.
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+
+Type RouteSpec struct {
+	// ...
+
+	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
+}
+```
+
+The reasoning for the field was mainly for future proofing the API to handle integrations
+with other third-party certificate management systems. But this API introduces inconsistency
+and does not fare well when the user wants to Ingress over directly creating routes.
+
+#### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields` 
+on Route to denote to the user which fields are managed by the secret-injector controller and 
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate` 
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update 
+  operation done by the user is only on the `.metadata.annotations`. 
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-reference`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+In scenarios where the user deletes the annotation, it is the controller's 
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e. 
```

Comment:
Assuming that the new controller will be watching the routes using a predicate on the new annotation. Will the controller receive an event when this annotation is removed?

Reply:
The new controller will watch all routes and process all routes in order to update `.status.certificateCondiitions`. The ones with the annotation will have additional logic to update `.spec.tls`. 

The only filter on the route informer is going to be excluding those with ingress owner references.

Reply:
I see, that's quite an overhead as we could use the new annotation as a predicate which (I guess) would enable the server side filtering instead of the client side (new controller). However that's maybe the only way around to clear `.spec.tls` fields after the annotation is removed.

Can you state this implementation detail in the EP please?

Reply:
Updated with the details.